Update the docs for private GKE. (#2190)
kubeflow/website: content/en/docs/gke/private-clusters.md

+++
title = "Securing Your Clusters"
description = "How to secure Kubeflow clusters using private GKE"
weight = 70
                    
+++

These instructions explain how to deploy Kubeflow using private GKE.

1. Follow the blueprint instructions to setup a management cluster

> 1. As a work around for (kubeflow/gcp-blueprints#32)[https://github.com/kubeflow/gcp-blueprints/issues/32]
>    modify the containercluster CRD schema in your management cluster to include missing fields

Reviewer: Suggestions:

- "work around" -> "workaround" or "work-around"?
- For links: "[text](link)"
- Add a short description of the GitHub issue and provide a definition for CRD (CustomResourceDefinition(s) - one word, according to k8s docs):
  - "As a workaround for Issue (...#32)[...32] (in CNRM 1.9.1, the [CustomResourceDefinition (CRD)](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#customresourcedefinitions) for container cluster is missing `ipAllocationPolicy` fields needed to create a private GKE cluster)..."
    - "CustomResourceDefinitions" is one word on [k8s docs](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#customresourcedefinitions) and three words on [GCP docs](https://cloud.google.com/migrate/anthos/docs/crd-overview).
- "containercluster" -> "container cluster"? (I couldn't find `containercluster` (one word) on Google Cloud docs unless it was a field/var name in some CLI output.)

For instance:

```suggestion
1. As a workaround for Issue
   [kubeflow/gcp-blueprints#32](https://github.com/kubeflow/gcp-blueprints/issues/32)
   (in CNRM 1.9.1, the [CustomResourceDefinition
   (CRD)](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#customresourcedefinitions)
   for container cluster is missing `ipAllocationPolicy` fields needed to create
   a private GKE cluster), modify the container cluster CRD schema in your
   management cluster to include the missing fields.
```

>    * See directions in that issue.

Reviewer: (Now I understand - you're referring to the directions in Issue 32 - got it!)

1. Fetch the blueprint

   ```
   kpt pkg get https://github.com/kubeflow/gcp-blueprints.git/kubeflow@master ./${PKGDIR}

>    ```

Reviewer: Suggestion: 

- Add: "where `{PKGDIR}` is the directory (for Kubeflow that you created earlier)?

Author: Added text this directory gets created.

> 1. Change to the kubeflow directory

Reviewer: "1. Change to the Kubeflow directory" (<- "kubeflow")

   ```
   cd ${PKGDIR}
   ```

1. Fetch Kubeflow manifests

   ```
   make get-pkg
   ```

1. Add the private GKE patches to your kustomization

   1. Open `instance/gcp_config`

>    1. In patchesStrategicMerge add 

Reviewer: ```
"   1. In `patchesStrategicMerge`, add 
```

      ```
      - ../../upstream/manifests/gcp/v2/privateGKE/cluster-private-patch.yaml
      ```

>    1. In resources add

Reviewer: ```
   1. In `resources`, add
```

      ```
      - ../../upstream/manifests/gcp/v2/privateGKE/
      ```

>    * Do not use `kustomize edit` to perform the above actions until [kubernetes-sigs/kustomize#2310](https://github.com/kubernetes-sigs/kustomize/issues/2310) is fixed

Reviewer: ```suggestion
   *Note:* Do not use `kustomize edit` to perform the above actions until Issue [kubernetes-sigs/kustomize#2310](https://github.com/kubernetes-sigs/kustomize/issues/2310) is fixed.
```

1. Open the `Makefile` and edit the `set-values` rule to invoke `kpt cfg set` with the desired values for

>    your deployment

Reviewer: (Lines 53-57 and elsewhere: Add full stops at the end of sentences)

   * Change `kpt cfg set ./instance gke.private false` to `kpt cfg set ./instance gke.private true`
   * You need to set region, location and zone because the deployment is a mix of zonal and regional resources and some which could be either

### Deploy Kubeflow

1. Configure the setters

   ```
   make set-values
   ```

1. Set environment variables with OAuth Client ID and Secret for IAP

   ```
   export CLIENT_ID=<client id>
   export CLIENT_SECRET=<client secret>
   ```

1. Deploy Kubeflow

   ```
   make apply
   ```

   * If this times out waiting for the cluster to be ready check if the reason the update failed is

>      because of [kubeflow/gcp-blueprints#34](https://github.com/kubeflow/gcp-blueprints/issues/35)

Author: Done.

  * If you want to completely disable all outbound internet access you will have to find some alternative solution
    to keep the JWKs in sync with your ISTIO policy

## Troubleshooting

* Cluster is stuck in provisioning state

  * Use the UI or gcloud to figure out what state the cluster is stuck in
  * If you use gcloud you need to look at the operation e.g.

    
    1. Find the operations
    
       ```
       gcloud --project=${PROJECT} container operations list
       ```

    1. Get operation details

       ```
       gcloud --project=${PROJECT} container operations describe --region=${REGION} ${OPERATION}
       ```

* Cluster health checks are failing.

   * This is usually because the firewall rules allowing the GKE health checks are not configured correctly

   * A good place to start is verifying they were created correctly

     ```
     kubectl --context=${MGMTCTXT} describe computefirewall
     ```

   * Turn on firewall rule logging to see what traffic is being blocked

     ```
     kpt cfg set ./upstream/manifests/gcp/v2/privateGKE/ log-firewalls true
     make apply
     ```

>    * To look for traffic blocked by firewall rules in stackdriver use a filter like the following

Reviewer: "Stackdriver"

      ```
      logName: "projects/${PROJECT}/logs/compute.googleapis.com%2Ffirewall" 
      jsonPayload.disposition = "DENIED"
      ```

      * **Logging must be enabled** on your firewall rules. You can enable it by using a kpt setter

        ```
        kpt cfg set ./upstream/manifests/gcp/v2/privateGKE/ log-firewalls true 
        ```

      * Change project to your project

      * Then look at the fields `jsonPayload.connection` this will tell you source and destination ips
      * Based on the IPs try to figure out where the traffic is coming from (e.g. node to master) and
        then match to appropriate firewall rules

      * For example

         ```
         connection: {
          dest_ip: "172.16.0.34"    
          dest_port: 443    
          protocol: 6    
          src_ip: "10.10.10.31"    
          src_port: 60556    
         }
         disposition: "DENIED" 
         ```

         * The destination IP in this case is for a GKE master so the firewall rules are not configured to correctly allow
           traffic to the master.

> * Common cause for networking related issue is is that some of the network resources (e.g. the Network, Routes, Firewall Rules, etc... ) don't get created

Reviewer: ```suggestion
* Common cause for networking related issue is that some of the network resources (e.g. the network, routes, firewall rules, etc... ) don't get created.
```

     * This could be because a reference is incorrect (e.g. firewall rules reference the wrong network)

>      * You can sanity check resources by doing kubectl describe and looking for errors.

Reviewer: (less slang)

```suggestion
     * You can double check the resources by running `kubectl describe` and looking for errors.
```

       * [kubeflow/gcp-blueprints#38](https://github.com/kubeflow/gcp-blueprints/issues/38) is tracking
          tools to automate this

> * GCR images can't be pulled

Reviewer: (Let's add a definition to "GCR", since it could be Google Cloud resources/Run/registry (depending on who's reading this))

  * This likely indicates an issue with access to private GCR; this could be an issue with

>     * DNS configurations - Check that the `DNSRecordSet` and `DNSManagedZone` CNRM resources are in a ready state

Reviewer 2: Sounds like it'll be better if we wait for all CNRM resources before continuing to apply other manifests.
Just a side note

Author: Ack.

>     * DNS configurations - Check that the `DNSRecordSet` and `DNSManagedZone` CNRM resources are in a ready state

Reviewer: ```suggestion
    * DNS configurations: Check that the `DNSRecordSet` and `DNSManagedZone` CNRM resources are in a ready state.
```

    * Routes - Ensure any default route to the internet has a larger value for the priority 

>         then any routes to private GCP APIs so that the private routes match first.

Reviewer: "Google CloudAPIs" (rebranding?)

>         * If image pull errors show IP addresses not the restricted.googleapis.com VIP then you have

Reviewer: ```suggestion
        * If image pull errors show IP addresses and not the `restricted.googleapis.com` VIP, then you have
```

          an issue with networking

    * Firewall rules

> * Access to whitelisted (non Google) sites is blocked

Reviewer: ```suggestion
* Access to whitelisted (non-Google) sites is blocked
```

  * The configuration uses CloudNat to allow selective access to sites

    * In addition to allowing IAP, this allows sites like GitHub to be whitelisted.

  * In order for CloudNat to work you need

    * A default route to the internet
    * Firewall rules to allow egress traffic to whitelisted sites

      * These rules need to be higher priority then the deny all firewall egress rules.

### Kubernetes Webhooks are blocked by firewall rules

A common failure mode is that webhooks for custom resources are blocked by default firewall rules.

> As explained in the [GKE Docs](https://cloud.google.com/kubernetes-engine/docs/how-to/private-clusters#add_firewall_rules) only connections from master to ports 443 and 10250

Reviewer: ```suggestion
As explained in the [GKE docs](https://cloud.google.com/kubernetes-engine/docs/how-to/private-clusters#add_firewall_rules), only connections from master to ports 443 and 10250
```

are allowed by default. If you have a webhook serving on a different port
you will need to add an explict ingress firewall rule to allow that port to be accessed.

These errors usually manifest as failures to create custom resources that depend on webhooks. An example

> error is

Reviewer: ```suggestion
error is:
```

Author: Resolved.

```
Error from server (InternalError): error when creating ".build/kubeflow-apps/cert-manager.io_v1alpha2_certificate_admission-webhook-cert.yaml": Internal error occurred: failed calling webhook "webhook.cert-manager.io": the server is currently unable to handle the request

> ```

Reviewer: Line 241: since there is already one troubleshooting section for KF on private GKE:

```diff
- * [Troubleshoot](/docs/gke/troubleshooting-gke) any issues you may
  find.
+ * [Troubleshoot](/docs/gke/troubleshooting-gke) any GKE issues you may
  find.
```

+++

> title = "Securing Your Clusters"

Reviewer: Hey @jlewi, I ran "private GKE" in the search bar:

<img width="800" alt="image" src="https://user-images.githubusercontent.com/19637339/93804335-104ee800-fc3e-11ea-82ac-e4e9bfaadca8.png">

I think maybe we could improve the UX by amending the title a little bit.

```diff
- title = "Securing Your Clusters"
+ title = "Securing Your Clusters with Private GKE"
description = "How to secure Kubeflow clusters using private GKE"
```

or

```diff
- title = "Securing Your Clusters"
+ title = "Securing Your Clusters on Private GKE"
description = "How to secure Kubeflow clusters using private GKE"
```

It would be similar to:

```
title = "Kubeflow On-premises on Anthos GKE"
```

LMKWYT

Reviewer 2: Sounds good to me. I'll merge this PR first because the original one is out of date

description = "How to secure Kubeflow clusters using private GKE"
weight = 70
                    
+++

These instructions explain how to deploy Kubeflow using private GKE.

> 1. Follow the blueprint instructions to setup a management cluster

Reviewer: Line 10: The steps below `1. Follow the blueprint instructions to setup a management cluster` - should be be indented to the right?

1. As a workaround for Issue
   [kubeflow/gcp-blueprints#32](https://github.com/kubeflow/gcp-blueprints/issues/32)
   (in CNRM 1.9.1, the [CustomResourceDefinition
   (CRD)](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#customresourcedefinitions)
   for container cluster is missing `ipAllocationPolicy` fields needed to create
   a private GKE cluster), modify the container cluster CRD schema in your
   management cluster to include the missing fields.

>    * See [kubeflow/gcp-blueprints#32](https://github.com/kubeflow/gcp-blueprints/issues/32)
>      to know whether this issue has been resolved in later versions of CNRM and if not
>      directions for how to work around it.

Reviewer: ```suggestion
   * Check Issue [kubeflow/gcp-blueprints#32](https://github.com/kubeflow/gcp-blueprints/issues/32)
     to find out if it has been resolved in later versions of CNRM. If the issue hasn't been resolved, there should be directions for a workaround.
```

(or something along these lines)

> 1. Fetch the blueprint

Reviewer: ```suggestion
1. Fetch the blueprint by running this command line:
```

   ```
   kpt pkg get https://github.com/kubeflow/gcp-blueprints.git/kubeflow@master ./${PKGDIR}
   ```

>    * This will create the directory ${PKGDIR} with the blueprint.

Reviewer: ```suggestion
   * This will create the directory `${PKGDIR}` with the blueprint.
```

1. Change to the Kubeflow directory:

   ```
   cd ${PKGDIR}
   ```

> 1. Fetch Kubeflow manifests

Reviewer: ```suggestion
1. Fetch Kubeflow manifests:
```
